### PR TITLE
fix Issue 23618 - Right Shift equals expressions on unsigned shorts s…

### PR DESCRIPTION
**Pull request by WalterBright**
…hould be unsigned right shift

### CODE REVIEW

**Review by ibuclaw — DISMISSED** [review]
Don't merge this on master, this is affecting industry.

**Comment by WalterBright** [comment]
@ibuclaw ???

**Comment by WalterBright** [comment]
I don't understand what change you are requesting. This is listed as a blocker bug. The fix is isolated to dmd. You said that gdc and ldc don't have this issue.

**Comment by dkorpel** [comment]
The requested change is to target stable

**Comment by WalterBright** [comment]
1. this problem has been there for many years, with apparently nobody reporting it. I don't see a rush to put it into stable
2. it carries a small risk of someone relying on the old behavior. That should disqualify it from stable.

**Comment by WalterBright** [comment]
But you can put it in stable if you prefer. I just want it fixed.

**Review by dkorpel — APPROVED** [review]

File: compiler/src/dmd/backend/cgelem.d [hunk 1]
```diff
@@ -483,8 +483,14 @@ private elem *fixconvop(elem *e)
              cop == OPu8_16 || cop == OPs8_16))
     {   if (e.Eoper != OPshlass && e.Eoper != OPshrass && e.Eoper != OPashrass)
             e.EV.E2 = el_una(icop,tym,e2);
-        //printf("after1\n");
-        //elem_print(e);
+
+        version (MARS)
```
[the remark this answers is not in the document]
Comment (dkorpel) [reply compiler/src/dmd/backend/cgelem.d:487]: Digital Mars C uses the backend as well: https://github.com/DigitalMars/Compiler/

I don't know if this check is needed, but I assume Walter had a reason to include it.

**Comment by ibuclaw** [comment]
> But you can put it in stable if you prefer. I just want it fixed.

Putting things in master won't fix the problem, you're just delaying the fix being released until 3 months down the line.  In the meantime companies are complaining and want it released _now_.

File: compiler/src/dmd/backend/cgelem.d (hunk 1 above)
[the remark this answers is not in the document]
Comment (WalterBright) [reply compiler/src/dmd/backend/cgelem.d:487]: The Digital Mars C/C++ compiler. They handle >> differently.

**Comment by WalterBright** [comment]
Thanks all for taking care of this.

**Comment by CyberShadow** [comment]
"Companies want [breaking change] right now" is perhaps not a good scalable approach in general.

Not that there isn't the problem that different users of D will want different proportions of stability vs. hotfixes (which suit them in that moment). I know Digger isn't amazing right now but I think having tooling which allows building site-specific variants of D is the best general solution to these problems.

**Comment by ibuclaw** [comment]
> "Companies want [breaking change] right now" is perhaps not a good scalable approach in general.
> 

Both GDC and LDC produce the correct code, as well as any C/C++ compiler (apart from DMC it seems, if I understand the `version(MARS)` condition right). DMD is just plain wrong here.

**Pull request merged**